streams/src/test/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStoreTest.java

```diff
@@ -124,18 +141,19 @@ public void before() {
             Serdes.String()
         );
         metrics.config().recordLevel(Sensor.RecordingLevel.DEBUG);
-        expect(context.metrics())
-            .andReturn(new StreamsMetricsImpl(metrics, "test", builtInMetricsVersion)).anyTimes();
-        expect(context.taskId()).andReturn(taskId).anyTimes();
-        expect(inner.name()).andReturn("metered").anyTimes();
+        expect(context.applicationId()).andStubReturn(APPLICATION_ID);
+        expect(context.metrics()).andStubReturn(new StreamsMetricsImpl(metrics, "test", builtInMetricsVersion));
+        expect(context.taskId()).andStubReturn(taskId);
+        expect(context.changelogFor(STORE_NAME)).andStubReturn(CHANGELOG_TOPIC);
```

Comment:
Unless we're actually verifying these interactions, we should probably use the MockInternalProcessorContext instead, and save ourselves from having to tweak these expectations every time we change the implementation details.

Reply:
I guess there is a misunderstanding. You cannot verify an expectation that is specified with `andStubReturn()`. That is the reason why I changed `andReturn().anyTimes()` with `andStubReturn()`. This also means that these expectations do not depend on the implementation details (i.e. how often the methods are called). The combination of nice mock with `andStubReturn()` has similar effect as using `MockInternalProcessorContext`. Refactoring the tests would unreasonably increase this PR. Hence, I would like not to do it in this PR.

Reply:
I think Bruno needs to give us all a short lesson on the correct usage of EasyMock. He explained the `StubReturn` thing to me on another PR a while back and I've been (slowly) trying to help migrate all the `.andReturn.anyTimes` usages over to this where appropriate (which is most places). It's definitely helped reduce the large number of EasyMock'ed tests that have to be fixed after every minor implementation change.

That said, there does seem to be one critical difference between using `.andStubReturn` and the actual `MockInternalProcessorContext`. If we add a new method to the `InternalProcessorContext` interface, for example, we then have to add this expectation to every test that calls it with an EasyMock context. Having had to do this a number of times, it's definitely a huge timesuck.

But I also agree that it doesn't need to be done as part of this PR. Maybe once the MockInternal and InternalMock processor contexts are finally merged we can do some reasonable cleanup of the tests

Reply:
Thanks, all. What Sophie mentioned is more in line with what I had in mind. The advantage of something like "MIPC" is that it just generally satisfies the API contract while not really doing stuff like writing to the brokers, etc. Because it's a contract-valid implementation of InternalProcessorContext, we can use it transparently to support any use case where we really don't care exactly what calls the unit under test makes into the context.

EasyMock is very powerful, but it can't magically implement a complex interface contract, so it can never fill this gap. On the other hand, if there's really only going to be one or two calls, then it's fine. And, of course, if we _need_ to verify interactions, exceptions, etc., then EasyMock gives us a very powerful set of utilities for it.

Anyway, we don't have to change this right now, but I've certainly spent many hours fiddling around with easymock expectations just like this, and it provides no value if you just want the thing to behave the way that MIPC already behaves. Which is why I pointed it out.

Reply:
I agree that EasyMock is not able to magically implement a complex interface contract. However, as I said earlier it would significantly increase the size of this PR to swap to `InternalMockProcessorContext` or one of its siblings right now.

Reply:
Ok, no need to change it now, then. Thanks for the discussion.